src/main/java/dev/lu15/voicechat/VoiceChatImpl.java

```diff
@@ -116,31 +131,124 @@ private void handle(@NotNull Player player, @NotNull HandshakePacket packet) {
             player.sendPacket(this.packetHandler.write(new HandshakeAcknowledgePacket(
                     event.getSecret(),
                     this.port,
-                    player.getUuid(), // why is this sent? the client already knows the player's uuid
-                    Codec.VOIP, // todo: configurable
-                    1024, // todo: configurable
-                    48, // todo: configurable
-                    1000, // todo: configurable
-                    false, // todo: configurable
+                    player.getUuid(),
+                    codec,
+                    mtu,
+                    distance,
+                    keepalive,
+                    groups,
                     this.publicAddress,
-                    false // todo: configurable
+                    recording // todo: configurable (recording)
             )));
+            groupManager.groups.forEach((id, group) -> {
+                player.sendPacket(this.packetHandler.write(new GroupCreatedPacket(group)));
+            });
         });
     }
 
     private void handle(@NotNull Player player, @NotNull UpdateStatePacket packet) {
         // todo: set state when players disconnect from voice chat server - NOT when they disconnect from the minecraft server
+        VoiceState oldstate = player.getTag(Tags.PLAYER_STATE);
+        UUID group = null;
+        if(oldstate!=null&&oldstate.group()!=null) {
+            group = oldstate.group();
+        }
         VoiceState state = new VoiceState(
                 packet.disabled(),
                 false,
                 player.getUuid(),
                 player.getUsername(),
+                group
+        );
+        player.setTag(Tags.PLAYER_STATE, state);
+        PacketSendingUtils.broadcastPlayPacket(this.packetHandler.write(new VoiceStatePacket(state)));
+        EventDispatcher.call(new PlayerUpdateVoiceStateEvent(player, state));
+    }
+
+    private void handle(@NotNull Player player, @NotNull CreateGroupPacket packet) {
+        if(player.getTag(Tags.PLAYER_STATE).disabled()) return;
+        boolean password = false;
+        if(packet.name().length()>24) return;
+        if(packet.password()!=null) {
+            if(packet.password().length()>24) return;
+            password = true;
+        }
+        Group group = new Group(UUID.randomUUID(), packet.name(), password, false, false, packet.type());
+        groupManager.groupPassword.put(group.id(), packet.password());
+        groupManager.groups.put(group.id(), group);
+        groupManager.playerGroups.put(player, group.id());
+        groupManager.groupPlayers.put(group.id(), new ArrayList<>(List.of(player)));
```

Comment:
This functionality should be placed inside `GroupManager`-owned methods, rather than being implemented in the packet processing.